### PR TITLE
Initial work towards native css support

### PR DESCRIPTION
**Pull request by sokra**
**What kind of change does this PR introduce?**
experiment

**Did you add tests for your changes?**
yes

**Does this PR introduce a breaking change?**
no

**What needs to be documented once your changes are merged?**
* `experiments.css: true` enables css support. See https://github.com/webpack/webpack/issues/14893 for progress and documentation.

### CODE REVIEW

**Comment by alexander-akait** [comment]
CSS allow to write `url` like:
```css
div {
  background: url('./img\
\
\
\
-other-part.png');
}
```

We normalize it in `css-loader`.

Here example of most of all `urls` - https://github.com/webpack-contrib/css-loader/blob/master/test/fixtures/url/url.css, we can add them here (all works in `css-loader`).

Also CSS allows to have weird things in `url`:
```css
div {  background: url(./img\'img.png); }
```

Also case for `@import` https://github.com/webpack-contrib/css-loader/blob/master/test/fixtures/import/import.css, but some of them broken for snapshots testing, broken things do not changed and keeping as is in CSS without resolving.

Also CSS allow to do circular imports (they just ignored because already loaded):

**file.css**

```css
@import "file.css";
```

**Comment by sokra** [comment]
Thanks. I'll add these as tests.

Some of them are a bit weird: `image-set("./img1x.png" 1x);`
Seem like using `url()` is optional here... That makes it a bit more complicated...

**Comment by sokra** [comment]
And I found another fancy thing in the spec:

```css
background-image: image('sprites.svg#xywh=40,0,20,20', black)
```

**Comment by alexander-akait** [comment]
> background-image: image('sprites.svg#xywh=40,0,20,20', black)

Yep, but in real world I have not seen anyone use this, `url` is shorter :smile:

**Comment by sokra** [comment]
> Yep, but in real world I have not seen anyone use this

It seem to be new and not yet supported:

image

It can show a fallback colors and crop images, so that's capability that url() doesn't have.

File: lib/asset/RawDataUrlModule.js [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+		super.updateHash(hash, context);
+	}
+
+	serialize(context) {
+		const { write } = context;
+
+		write(this.url);
+		write(this.identifierStr);
+		write(this.readableIdentifierStr);
+		write(this.runtimeRequirements);
```
Comment (vankop) [inline lib/asset/RawDataUrlModule.js:128]: there is no usage of `this.runtimeRequirements` maybe need a small refactoring of `this. codeGeneration` method

File: lib/css/CssGenerator.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+		}
+
+		return originalSource.size();
+	}
+
+	/**
+	 * @param {Hash} hash hash that will be modified
+	 * @param {UpdateHashContext} updateHashContext context for updating hash
+	 */
+	updateHash(hash, { module }) {}
```
Comment (vankop) [inline lib/css/CssGenerator.js:92]: should we do here same as https://github.com/webpack/webpack/pull/14857

File: lib/css/CssLoadingRuntimeModule.js [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+				: "",
+			"link.setAttribute(loadingAttribute, 1);",
+			'link.rel = "stylesheet";',
+			withHmr ? 'if(hmr) link.media = "print and screen";' : "",
+			"link.href = url;",
+			crossOriginLoading
+				? Template.asString([
+						"if (link.src.indexOf(window.location.origin + '/') !== 0) {",
+						Template.indent(
+							`link.crossOrigin = ${JSON.stringify(crossOriginLoading)};`
```
Comment (vankop) [inline lib/css/CssLoadingRuntimeModule.js:116]: interesting small improvement https://github.com/webpack/webpack/pull/10064#issuecomment-560369924 possible here

File: lib/css/CssModulesPlugin.js [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+					.tap(plugin, handler);
+				compilation.hooks.runtimeRequirementInTree
+					.for(RuntimeGlobals.hmrDownloadUpdateHandlers)
+					.tap(plugin, handler);
+			}
+		);
+	}
+
+	getOrderedChunkCssModules(chunk, chunkGraph) {
+		const cssImports = chunkGraph.getOrderedChunkModulesIterableBySourceType(
```
Comment (vankop) [inline lib/css/CssModulesPlugin.js:189]: probably `css-import` and `css` should be in one "order" call. e.g. in code below
```
import "https://h.ost/a.css" // css-import
import "./a.css"
```
hash order should be ['css-import', 'css']
and next one
```
import "./a.css"
import "https://h.ost/a.css" // css-import
```
hash order should be ['css', 'css-import']

maybe `getOrderedChunkModulesIterableBySourceType ` could be improved to allow pass `Set` as argument

File: lib/css/CssParser.js [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+					pos++;
+					continue;
+				}
+				if (cc === 10) pos++;
+				break;
+			}
+			return pos;
+		};
+		walkCssTokens(source, {
+			url: (input, start, end, contentStart, contentEnd) => {
```
Comment (vankop) [inline lib/css/CssParser.js:128]: maybe we need a hooks here (at least in webpack@6)

File: lib/css/CssModulesPlugin.js (hunk 4 above)
Comment (sokra) [reply lib/css/CssModulesPlugin.js:189]: In a CSS file all `@import` must be before normal rules. So this enforces this order.

**Pull request merged**